<a href="https://colab.research.google.com/github/LazarevaL/VK_intern_ML/blob/main/VK_intern_ML.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Загрузка и обработка данных

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('intern_task.csv')

In [3]:
df

rank  query_id  feature_0  feature_1  feature_2  feature_3  feature_4  \
0         0        10        1.0        0.0        1.0        3.0        3.0   
1         1        10        3.0        0.0        3.0        0.0        3.0   
2         0        10        3.0        0.0        2.0        0.0        3.0   
3         1        10        3.0        0.0        3.0        0.0        3.0   
4         2        10        3.0        0.0        3.0        1.0        3.0   
...     ...       ...        ...        ...        ...        ...        ...   
76510     1     10135        2.0        0.0        0.0        1.0        2.0   
76511     1     10135        2.0        0.0        1.0        1.0        2.0   
76512     1     10135        2.0        0.0        2.0        1.0        2.0   
76513     1     10135        2.0        0.0        2.0        1.0        2.0   
76514     1     10135        2.0        0.0        2.0        0.0        2.0   

       feature_5  feature_6  feature_7  ...  feature_134  feature_135  \
0       0.333333        0.0   0.333333  ...          0.0     0.000000   
1       1.000000        0.0   1.000000  ...          0.0     0.000000   
2       1.000000        0.0   0.666667  ...          0.0     0.000000   
3       1.000000        0.0   1.000000  ...          0.0     0.000000   
4       1.000000        0.0   1.000000  ...        273.0    79.670665   
...          ...        ...        ...  ...          ...          ...   
76510   1.000000        0.0   0.000000  ...          0.0     0.000000   
76511   1.000000        0.0   0.500000  ...          0.0     0.000000   
76512   1.000000        0.0   1.000000  ...          5.0    91.000000   
76513   1.000000        0.0   1.000000  ...          0.0     0.000000   
76514   1.000000        0.0   1.000000  ...          NaN          NaN   

       feature_136  feature_137  feature_138  feature_139  feature_140  \
0         0.454545     0.890238     8.655534     1.000000     0.077778   
1         0.000000     0.773976    23.130514     0.000000     0.027826   
2         0.000000     0.918308    13.351339     0.000000     0.014925   
3         0.000000     0.975355    18.240926     0.000000     0.053140   
4         0.200000     0.990119    31.786048     0.333333     0.046512   
...            ...          ...          ...          ...          ...   
76510     0.142857     0.999639     0.000000     0.500000     0.043590   
76511     0.125000     0.974958     6.691363     0.500000     0.029183   
76512     0.125000     0.972117    12.598439     0.500000     0.020376   
76513     0.111111     1.000000    12.598439     0.500000     0.003429   
76514          NaN          NaN          NaN          NaN          NaN   

       feature_141  feature_142  feature_143  
0         0.002222          1.0     0.333333  
1         0.000430         44.0    14.666667  
2         0.000104         22.0     7.333333  
3         0.000255          8.0     2.666667  
4         0.000307         24.0     8.000000  
...            ...          ...          ...  
76510     0.000002         33.0    16.500000  
76511     0.000065         28.0    14.000000  
76512     0.000032         23.0    11.500000  
76513     0.000000          8.0     4.000000  
76514          NaN          NaN          NaN  

[76515 rows x 146 columns]

In [4]:
df['rank'].value_counts()

rank
0    42805
1    22135
2     9905
3     1217
4      453
Name: count, dtype: int64

In [6]:
df.describe()

rank      query_id     feature_0     feature_1     feature_2  \
count  76515.000000  76515.000000  76515.000000  76515.000000  76515.000000   
mean       0.619591   5252.164674      1.968438      0.219186      1.203581   
std        0.813073   2999.691648      1.222736      0.601284      1.056596   
min        0.000000     10.000000      0.000000      0.000000      0.000000   
25%        0.000000   2650.000000      1.000000      0.000000      0.000000   
50%        0.000000   5305.000000      2.000000      0.000000      1.000000   
75%        1.000000   8035.000000      3.000000      0.000000      2.000000   
max        4.000000  10135.000000      8.000000      6.000000      6.000000   

          feature_3     feature_4     feature_5     feature_6     feature_7  \
count  76515.000000  76515.000000  76515.000000  76515.000000  76515.000000   
mean       0.537777      2.010887      0.804078      0.099861      0.526575   
std        0.800674      1.192559      0.341939      0.266116      0.417180   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      1.000000      0.666667      0.000000      0.000000   
50%        0.000000      2.000000      1.000000      0.000000      0.500000   
75%        1.000000      3.000000      1.000000      0.000000      1.000000   
max        6.000000      8.000000      1.000000      1.000000      1.000000   

       ...   feature_134   feature_135   feature_136   feature_137  \
count  ...  7.651400e+04  7.651400e+04  76514.000000  76514.000000   
mean   ...  6.071034e+02  5.238281e+03      0.068273      0.740544   
std    ...  2.281264e+04  1.446071e+06      0.101253      0.347901   
min    ...  0.000000e+00  0.000000e+00      0.000000      0.000000   
25%    ...  0.000000e+00  0.000000e+00      0.000000      0.623210   
50%    ...  0.000000e+00  0.000000e+00      0.000000      0.917337   
75%    ...  1.000000e+00  1.000000e+00      0.125000      0.996810   
max    ...  1.731553e+06  4.000000e+08      0.818182      1.000000   

        feature_138   feature_139   feature_140   feature_141   feature_142  \
count  76514.000000  76514.000000  76514.000000  76514.000000  76514.000000   
mean      10.973229      0.260314      0.025886      0.000294     22.608764   
std       13.501283      0.377056      0.031101      0.002324     35.203027   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.006375      0.000000      4.000000   
50%        9.304890      0.000000      0.017333      0.000012     12.000000   
75%       16.415114      0.500000      0.033868      0.000125     28.000000   
max     1043.726592      1.000000      0.617647      0.250000   1984.000000   

        feature_143  
count  76514.000000  
mean       9.419858  
std       13.441573  
min        0.000000  
25%        2.000000  
50%        5.500000  
75%       12.000000  
max      496.000000  

[8 rows x 146 columns]

In [7]:
df.isnull().sum()

rank           0
query_id       0
feature_0      0
feature_1      0
feature_2      0
              ..
feature_139    1
feature_140    1
feature_141    1
feature_142    1
feature_143    1
Length: 146, dtype: int64

In [8]:
#поскольку в данных почти нет/очень мало пропусков, мы можем дропнуть ряды с nan, поскольку данные от этого не пострадают
df.dropna(inplace = True)

В общем, датасет уже предобработан и готов к построению моделей ранжирования: отсутсвующие значения удалены, все фичи числовые и не требуют преобразования, проверить наличие дубликатов  в таких данных не представляется возможных, а в визуализация данных нет смысла поскольку никакой новой информации это не даст.

##XGBoost

In [9]:
import xgboost as xgb
from sklearn.metrics import ndcg_score
from sklearn.model_selection import train_test_split
from sklearn.metrics import top_k_accuracy_score

In [10]:
train, test = train_test_split(df, test_size=0.3, random_state=42)

In [11]:
train_df = train.sort_values(by=['query_id']).reset_index(drop=True)
test_df = test.sort_values(by=['query_id']).reset_index(drop=True)

In [12]:
X_train = train_df.drop(columns=['rank', 'query_id'])
y_train = train_df['rank']
qid_train = train_df['query_id']

X_test = test_df.drop(columns=['rank', 'query_id'])
y_test = test_df['rank']
qid_test = test_df['query_id']

In [13]:
X_train

feature_0  feature_1  feature_2  feature_3  feature_4  feature_5  \
0            3.0        3.0        3.0        1.0        3.0   1.000000   
1            0.0        0.0        0.0        0.0        0.0   0.000000   
2            2.0        0.0        1.0        0.0        2.0   0.666667   
3            2.0        0.0        2.0        1.0        2.0   0.666667   
4            3.0        0.0        1.0        0.0        3.0   1.000000   
...          ...        ...        ...        ...        ...        ...   
53554        2.0        0.0        2.0        1.0        2.0   1.000000   
53555        0.0        0.0        0.0        1.0        1.0   0.000000   
53556        2.0        0.0        1.0        0.0        2.0   1.000000   
53557        2.0        0.0        2.0        0.0        2.0   1.000000   
53558        2.0        0.0        2.0        0.0        2.0   1.000000   

       feature_6  feature_7   feature_8  feature_9  ...  feature_134  \
0            1.0   1.000000    0.382560   1.000000  ...          1.0   
1            0.0   0.000000    0.153024   0.000000  ...          0.0   
2            0.0   0.333333    0.034005   0.666667  ...        218.0   
3            0.0   0.666667    0.378310   0.666667  ...          0.0   
4            0.0   0.333333    0.331552   1.000000  ...          0.0   
...          ...        ...         ...        ...  ...          ...   
53554        0.0   1.000000  324.925507   1.000000  ...          0.0   
53555        0.0   0.000000  324.738477   0.500000  ...          0.0   
53556        0.0   0.500000  324.942510   1.000000  ...          0.0   
53557        0.0   1.000000  324.976515   1.000000  ...         15.0   
53558        0.0   1.000000  324.759731   1.000000  ...          0.0   

       feature_135  feature_136  feature_137  feature_138  feature_139  \
0         1.800000     0.125000     0.989621    18.240926     0.333333   
1         0.000000     0.000000     0.000000     0.000000     0.000000   
2        55.069946     0.000000     0.448807     4.695805     0.000000   
3         0.000000     0.166667     0.907969    13.545122     0.333333   
4         0.000000     0.000000     0.931169     8.655534     0.000000   
...            ...          ...          ...          ...          ...   
53554     0.000000     0.111111     0.982735    12.598439     0.500000   
53555     0.000000     0.166667     0.000000     0.000000     0.500000   
53556     0.000000     0.000000     0.935562     6.691363     0.000000   
53557    22.236667     0.000000     1.000000    12.598439     0.000000   
53558     0.000000     0.000000     0.984269    19.289802     0.000000   

       feature_140  feature_141  feature_142  feature_143  
0         0.007376     0.000001         14.0     4.666667  
1         0.000000     0.000000          0.0     0.000000  
2         0.002153     0.000002          5.0     1.666667  
3         0.019259     0.000180         10.0     3.333333  
4         0.007549     0.000021         18.0     6.000000  
...            ...          ...          ...          ...  
53554     0.016900     0.000013         26.0    13.000000  
53555     0.083333     0.000000          0.0     0.000000  
53556     0.014493     0.000053         33.0    16.500000  
53557     0.007648     0.000000          6.0     3.000000  
53558     0.018651     0.000011         49.0    24.500000  

[53559 rows x 144 columns]

In [21]:
ranker = xgb.XGBRanker(
    max_depth =  8,
    objective = 'rank:ndcg',
    eval_metric=["ndcg"],
    seed = 42
)
ranker.fit(
    X_train,
    y_train,
    qid=qid_train,
    verbose=True
)

XGBRanker(base_score=None, booster=None, callbacks=None, colsample_bylevel=None,
          colsample_bynode=None, colsample_bytree=None, device=None,
          early_stopping_rounds=None, enable_categorical=False,
          eval_metric=['ndcg'], feature_types=None, gamma=None,
          grow_policy=None, importance_type=None, interaction_constraints=None,
          learning_rate=None, max_bin=None, max_cat_threshold=None,
          max_cat_to_onehot=None, max_delta_step=None, max_depth=8,
          max_leaves=None, min_child_weight=None, missing=nan,
          monotone_constraints=None, multi_strategy=None, n_estimators=None,
          n_jobs=None, num_parallel_tree=None, random_state=None, ...)

In [22]:
#метрика на трейне
y_pred_tr = ranker.predict(X_train)
ndcg_score([y_train], [y_pred_tr])

0.9631863514361698

In [23]:
#метрика на тесте
y_pred_test = ranker.predict(X_test)
ndcg_score([y_test], [y_pred_test])

0.9176603292285794

##CatBoost

In [25]:
#pip install catboost
import catboost as cat

In [31]:
train_pool = cat.Pool(
    data=X_train,
    label=y_train,
    group_id=qid_train
)

test_pool = cat.Pool(
    data=X_test,
    label=y_test,
    group_id=qid_test
)

In [34]:
ranker_ct = cat.CatBoostRanker(
    max_depth =  8,
    eval_metric="NDCG",
    random_seed = 42
)
ranker_ct.fit(
    train_pool,
    eval_set = test_pool
)

0:	test: 0.6945202	best: 0.6945202 (0)	total: 582ms	remaining: 9m 41s
1:	test: 0.7202455	best: 0.7202455 (1)	total: 1.1s	remaining: 9m 11s
2:	test: 0.7325912	best: 0.7325912 (2)	total: 1.61s	remaining: 8m 55s
3:	test: 0.7435083	best: 0.7435083 (3)	total: 2.15s	remaining: 8m 54s
4:	test: 0.7438560	best: 0.7438560 (4)	total: 2.52s	remaining: 8m 21s
5:	test: 0.7473093	best: 0.7473093 (5)	total: 2.78s	remaining: 7m 39s
6:	test: 0.7469394	best: 0.7473093 (5)	total: 3.03s	remaining: 7m 9s
7:	test: 0.7491578	best: 0.7491578 (7)	total: 3.3s	remaining: 6m 49s
8:	test: 0.7510980	best: 0.7510980 (8)	total: 3.54s	remaining: 6m 30s
9:	test: 0.7547973	best: 0.7547973 (9)	total: 3.8s	remaining: 6m 16s
10:	test: 0.7562967	best: 0.7562967 (10)	total: 4.05s	remaining: 6m 3s
11:	test: 0.7570590	best: 0.7570590 (11)	total: 4.32s	remaining: 5m 55s
12:	test: 0.7580486	best: 0.7580486 (12)	total: 4.56s	remaining: 5m 46s
13:	test: 0.7618719	best: 0.7618719 (13)	total: 4.81s	remaining: 5m 38s
14:	test: 0.76064

In [37]:
#метрика на трейне
y_pred_tr = ranker_ct.predict(X_train)
ndcg_score([y_train], [y_pred_tr])

0.9551741527820774

In [38]:
#метрика на тесте
y_pred_test = ranker_ct.predict(X_test)
ndcg_score([y_test], [y_pred_test])

0.9195688925376847

##Результат

При использовании двух библиотек - XGBoost и CatBoost, метрики получились одного порядка и очень близкие друг другу, поэтому однозначно сказать, какой из алгоритмов показал себя лучше невозможно.

|     | XGB | CatBoost |
| --- | --- | --- |
| train | 0.963186 | 0.955174 |
|test | 0.917660 | 0.919568 |